libraries/standard/mqtt/src/iot_mqtt_api.c

```diff
@@ -803,12 +803,22 @@ static IotMqttError_t _sendMqttMessage( _mqttOperation_t * pMqttOperation,
 {
     IotMqttError_t status = IOT_MQTT_SUCCESS;
 
+    /* Cached value of initialized operation status to check it after
+     * the operation has executed for the MQTT_INTERNAL_FLAG_BLOCK_ON_SEND case. */
+    IotMqttError_t cachedStatus = pMqttOperation->u.operation.status;
+
     IotMqtt_Assert( pMqttOperation != NULL );
 
     /* Send the SUBSCRIBE packet. */
     if( ( flags & MQTT_INTERNAL_FLAG_BLOCK_ON_SEND ) == MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )
     {
         _IotMqtt_ProcessSend( IOT_SYSTEM_TASKPOOL, pMqttOperation->job, pMqttOperation );
+
+        /* Update return status value if operation object status changed after job execution. */
+        if( pMqttOperation->u.operation.status != cachedStatus )
```

Comment:
This may be causing double free issues.  Please try moving code to check u.operation.status in PublishAsync() ( for QOS0 and MQTT_INTERNAL_FLAG_BLOCK_ON_SEND )

Reply:
Have used a different approach of not utilizing the complex "operations" infrastructure when dealing with the simplistic case of synchronous QoS 0 request.